src/python/pants/rules/core/binary.py

```diff
@@ -0,0 +1,68 @@
+# Copyright 2019 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from dataclasses import dataclass
+from typing import Any
+
+from pants.backend.python.rules.pex import Pex
+from pants.build_graph.address import Address
+from pants.engine.addressable import BuildFileAddresses
+from pants.engine.console import Console
+from pants.engine.fs import DirectoryToMaterialize, Workspace
+from pants.engine.goal import Goal, LineOriented
+from pants.engine.legacy.graph import HydratedTarget
+from pants.engine.legacy.structs import PythonBinaryAdaptor
+from pants.engine.rules import console_rule, optionable_rule, rule, union
+from pants.engine.selectors import Get
+
+
+@dataclass(frozen=True)
+class Binary(LineOriented, Goal):
+  name = 'binary'
+
+
+@union
+@dataclass(frozen=True)
+class BinaryTarget:
+  pass
+
+
+@union
+@dataclass(frozen=True)
+class CreatedBinary:
+  _inner: Any
+
+  def write_to_dist(self, print_stdout: Any, workspace: Workspace) -> None:
```

Comment:
This feels like it should be inlined into the `@console_rule`? Or was it broken out for testing or something?